source/slang/slang-ir-deduplicate.cpp

```diff
@@ -0,0 +1,80 @@
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+    struct DeduplicateContext
+    {
+        SharedIRBuilder* builder;
+        IRInst* addValue(IRInst* value)
+        {
+            if (!value) return nullptr;
+            if (as<IRType>(value))
+                return addTypeValue(value);
+            if (auto constValue = as<IRConstant>(value))
+                return addConstantValue(constValue);
+            return value;
+        }
+        IRInst* addConstantValue(IRConstant* value)
+        {
+            IRConstantKey key = { value };
+            if (auto newValue = builder->constantMap.TryGetValue(key))
+                return *newValue;
+            builder->constantMap[key] = value;
+            return value;
+        }
+        IRInst* addTypeValue(IRInst* value)
+        {
+            // Do not deduplicate struct types.
```

Comment:
This same exception should apply to interface types, right?

I feel like we've made these same distinctions in other passes, and we really need a way to identify structural-vs-nominal types that is shared across parts of the IR.

Reply:
I was not quite sure on interfaces, as it seems fine to deduplicate equivalent interface types?

Reply:
We don't deduplicate `struct`s because its fields are defined as children, not as operands. But `interface` types are using operands so they can deduplicate correctly.

Reply:
Well, we could switch the handling of `struct` types around to use operands now, but we'd still want to avoid applying deduplication to them.

I guess nothing would go drastically wrong right now if we deduplicate `interface` types, but I think it is wrong on principle, and we'd probably want to avoid potential problems down the line. E.g., if we start emitting RTTI into output binaries, then we'd want distinct interfaces to get distinct RTTI objects, even if they are structurally identical. This would be especially important if we starting allowing RTTI queries to check for things like the types that implement an interface, or the `extension`s that apply to one.